common/messaging/metrics_producer.go

```diff
@@ -26,38 +26,52 @@ import (
 	"github.com/uber/cadence/common/metrics"
 )
 
-type (
-	metricsProducer struct {
-		producer      Producer
-		metricsClient metrics.Client
+type MetricsProducer struct {
+	producer Producer
+	scope    metrics.Scope
+	tags     []metrics.Tag
+}
+
+type MetricProducerOptions func(*MetricsProducer)
+
+func WithMetricTags(tags ...metrics.Tag) MetricProducerOptions {
+	return func(p *MetricsProducer) {
+		p.tags = tags
 	}
-)
+}
 
 // NewMetricProducer creates a new instance of producer that emits metrics
 func NewMetricProducer(
 	producer Producer,
 	metricsClient metrics.Client,
+	opts ...MetricProducerOptions,
 ) Producer {
-	return &metricsProducer{
-		producer:      producer,
-		metricsClient: metricsClient,
+	p := &MetricsProducer{
+		producer: producer,
+	}
+
+	for _, opt := range opts {
+		opt(p)
 	}
+
+	p.scope = metricsClient.Scope(metrics.MessagingClientPublishScope, p.tags...)
```

Comment:
Why are we not initializing scope in &MetricsProducer while object creation?

Reply:
Because `p.tags` is initially empty and it gets populated after applying `opts` above.